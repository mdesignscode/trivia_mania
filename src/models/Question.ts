import { DataTypes, Model, type CreationOptional, type InferAttributes, type InferCreationAttributes } from 'sequelize';
import sequelize from 'utils/sequelize';


export default class Question extends Model<InferAttributes<Question>, InferCreationAttributes<Question>> {
        declare id: CreationOptional<number>;
        declare category: string;
        declare answers: string[];
        declare correctAnswer: string;
        declare question: string;
        declare difficulty: string;
};
export type TQuestionAttributes = InferAttributes<Question>;

Question.init(
        {
                id: {
                        type: DataTypes.INTEGER,
                        autoIncrement: true,
                        primaryKey: true,
                },
                category: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
                answers: {
                        type: DataTypes.JSON,
                        allowNull: false,
                },
                correctAnswer: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
                question: {
                        type: DataTypes.STRING,
                        allowNull: false,
                        unique: true,
                },
                difficulty: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
        }, {
        sequelize
}
);

