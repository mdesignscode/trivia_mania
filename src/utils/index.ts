import { browser } from '$app/environment';

export function decodeHTMLEntities(text: string) {
        if (browser) {
                const element = document.createElement("div");
                element.innerHTML = text;
                return element.textContent || "";
        }
}

export const colorMap: { [key: string]: string } = {
        easy: "green",
        medium: "gold",
        hard: "red",
};

export function capitalize(str: string): string {
	if (!str) return '';
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

