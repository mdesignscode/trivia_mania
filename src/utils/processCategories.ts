export type TQuestionData = {
        category: string;
        difficulty: string;
        count: number;
};
export type TAllCategories = Record<string, Record<string, TQuestionData>>;

export type TProcessResult = {
        allCategories: TAllCategories;
        allDifficulties: string[];
};

export function processCategories(
        categories: TQuestionData[],
        difficulties: string[]
): TProcessResult {
        const allCategories: TAllCategories = {};

        // map difficulties to categories
        categories.forEach(({ difficulty, count, category }) => {
                if (!allCategories[category]) allCategories[category] = {};
                allCategories[category][difficulty] = { difficulty, count, category };
        });

        // add "all difficulties" count for each category
        Object.keys(allCategories).forEach((key) => {
                const easy = allCategories[key].easy?.count || 0;
                const medium = allCategories[key].medium?.count || 0;
                const hard = allCategories[key].hard?.count || 0;
                const allCount = easy + medium + hard;

                allCategories[key]['all difficulties'] = {
                        difficulty: 'all difficulties',
                        count: allCount,
                        category: key
                };
        });

        // total "All Categories" aggregate
        const totalCategories: Record<string, TQuestionData> = {
                easy: { difficulty: 'easy', count: 0, category: 'All Categories' },
                medium: { difficulty: 'medium', count: 0, category: 'All Categories' },
                hard: { difficulty: 'hard', count: 0, category: 'All Categories' },
                'all difficulties': {
                        difficulty: 'all difficulties',
                        count: 0,
                        category: 'All Categories'
                }
        };

        Object.values(allCategories).forEach((catGroup) => {
                ['easy', 'medium', 'hard'].forEach((level) => {
                        if (catGroup[level]) {
                                totalCategories[level].count += catGroup[level].count;
                                totalCategories['all difficulties'].count += catGroup[level].count;
                        }
                });
        });

        allCategories['All Categories'] = totalCategories;

        return {
                allCategories,
                allDifficulties: [...difficulties]
        };
}

