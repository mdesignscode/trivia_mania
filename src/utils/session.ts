import { db } from '$lib/db';
import { encodeBase32LowerCaseNoPadding, encodeHexLowerCase } from '@oslojs/encoding';
import { sha256 } from '@oslojs/crypto/sha2';

export function generateSessionToken(): string {
        const bytes = new Uint8Array(20);
        crypto.getRandomValues(bytes);
        return encodeBase32LowerCaseNoPadding(bytes);
}

export function createSession(token: string, userId: number) {
        const sessionId = encodeHexLowerCase(sha256(new TextEncoder().encode(token)));
        const expiresAt = new Date(Date.now() + 1000 * 60 * 60 * 24 * 30); // 30 days

        db.prepare(
                'INSERT INTO session (id, user_id, expires_at) VALUES (?, ?, ?)'
        ).run(sessionId, userId, Math.floor(expiresAt.getTime() / 1000));
}

export function validateSessionToken(token: string) {
        const sessionId = encodeHexLowerCase(sha256(new TextEncoder().encode(token)));

        const row = db.prepare(`
    SELECT session.id, session.user_id, session.expires_at, user.username
    FROM session
    INNER JOIN user ON user.id = session.user_id
    WHERE session.id = ?
  `).get(sessionId);

        if (!row) {
                return { session: null, user: null };
        }

        const session = {
                id: row.id as string,
                userId: row.user_id as number,
                expiresAt: new Date(row.expires_at * 1000)
        };
        const user = {
                username: row.username as string
        };

        if (Date.now() >= session.expiresAt.getTime()) {
                db.prepare('DELETE FROM session WHERE id = ?').run(session.id);
                return { session: null, user: null };
        }

        return { session, user };
}

export function invalidateSession(sessionId: string) {
        db.prepare('DELETE FROM session WHERE id = ?').run(sessionId);
}

