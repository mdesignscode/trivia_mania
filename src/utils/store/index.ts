import globalStore from './global.svelte';
import { questionStore, playStore } from './playStore.svelte';

export {
        globalStore, playStore, questionStore,
}

