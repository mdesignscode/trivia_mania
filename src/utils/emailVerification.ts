import { env } from '$env/dynamic/private';
import { EmailVerification } from '../models';
import nodemailer from 'nodemailer';

const sender = env.EMAIL_ADDRESS;
const pass = env.EMAIL_CODE;

// Configure the transporter
const transporter = nodemailer.createTransport({
        host: 'smtp.gmail.com', // e.g. smtp.gmail.com
        port: 587,
        secure: false, // true for 465, false for other ports
        auth: {
                user: sender,
                pass,
        },
});

// Send email function
export const sendEmail = async ({ email, username }: { username: string, email: string }) => {
        // generate email verification code
        const code = await generateEmailVerificationCode(email);

        // email body
        const textBody = `<p>Hi ${username},</p>
    <p>Thank you for signing up with <strong>Trivia Mania</strong>! To complete your registration, please use the verification code below:</p>
    <div class="code-box">${code}</div>
    <p>This code will expire in <strong>10 minutes</strong>. If you did not request this, you can safely ignore this email.</p>
    <p>Need help? <a href="mailto:${sender}">Contact our support team</a>.</p>
    <div class="footer">
      &copy; 2025 Trivia Mania. All rights reserved.
      <br>
      <a href="https://triviamania.vercel.app">https://triviamania.vercel.app</a>
    </div>`

        try {
                // send email
                await transporter.sendMail({
                        from: '"Trivia Mania" ejue1vs7n@mozmail.com', // sender address
                        to: email, // list of receivers
                        subject: 'Verify Your Email Address', // Subject line
                        html: textBody, // HTML body
                });

                return { status: 'success', data: code };
        } catch (error) {
                console.error('Error sending email:', error);
                return { status: 'fail', data: error.message };
        }
};


export async function generateEmailVerificationCode(email: string): Promise<string> {
        const code = Math.floor(100000 + Math.random() * 900000).toString(); // 6-digit code
        const expiresAt = new Date(Date.now() + 10 * 60 * 1000); // 10 minutes from now

        // existing verification should be updated
        const existingVerification = await EmailVerification.findOne({
                where: { email },
        });

        if (existingVerification) {
                existingVerification.setDataValue('code', code);
                existingVerification.setDataValue('expiresAt', expiresAt);
                await existingVerification.save();
                return code;
        }

        await EmailVerification.create({
                email,
                code,
                expiresAt,
                used: false,
        });

        return code;
}

export async function validateEmailVerificationCode(email: string, inputCode: string): Promise<{ success: boolean; message: string }> {
        const record = await EmailVerification.findOne({
                where: {
                        email,
                },
        });

        const failedResponse = { success: false, message: 'Invalid or expired verification code.' };

        if (record) {
                // if code hasn't expired yet
                const codeExpired = record.getDataValue('expiresAt') < new Date();
                if (codeExpired) return failedResponse;

                // is code correct
                const code = record.getDataValue('code');
                if (code !== inputCode) return failedResponse;

                // verified
                record.setDataValue('used', true)
                await record.save();

                return { success: true, message: 'Verification successful.' };
        }

        // record not found
        return failedResponse;
}

