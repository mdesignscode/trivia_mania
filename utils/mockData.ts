import { CategoryStat, DifficultyStat, IUserStats } from "@/models/interfaces";
import { IQuestion } from "@/models/interfaces";

type TStat = Record<string, number>

export const mockInitialProgress: IUserStats = {
  total: {
    correctAnswered: 1,
    answered: 1,
  },
  easy: {
    answered: 1,
    correctAnswered: 1,
  } as DifficultyStat,
  Science: {
    easy: {
      answered: 1,
      correctAnswered: 1,
    } as DifficultyStat,
  } as CategoryStat,
};

// create mock questions
export const mockQuestions: Record<string, IQuestion> = {
  "63805aab-b589-4256-9493-769eecc23966": {
    "category": "Entertainment: Film",
    "answers": [
      "Batman v Superman: Superapocalypse",
      "Batman v Superman: Black of Knight",
      "Batman v Superman: Dawn of Justice",
      "Batman v Superman: Knightfall"
    ],
    "correctAnswer": "Batman v Superman: Dawn of Justice",
    "id": "63805aab-b589-4256-9493-769eecc23966",
    "question": "Which movie released in 2016 features Superman and Batman fighting?",
    "difficulty": "medium"
  },
  "36240604-80aa-4ddb-bb87-5ce345bfd379": {
    "category": "Entertainment: Film",
    "answers": ["R2-D2", "AA-A", "BB-8", "BB-3"],
    "correctAnswer": "BB-8",
    "id": "36240604-80aa-4ddb-bb87-5ce345bfd379",
    "question": "What is the orange and white bot&#039;s name in &quot;Star Wars: The Force Awakens&quot;?",
    "difficulty": "easy"
  },
  "12c4b40f-7fa9-405d-81de-b64c54a0caea": {
    "category": "Entertainment: Music",
    "answers": [
      "James Blunt - &quot;1973&quot;",
      "Prince - &quot;1999&quot;",
      "The Smashing Pumpkins - &quot;1979&quot;",
      "David Bowie - &quot;1984&quot;"
    ],
    "correctAnswer": "The Smashing Pumpkins - &quot;1979&quot;",
    "id": "12c4b40f-7fa9-405d-81de-b64c54a0caea",
    "question": "Which of these songs was released in 1996?",
    "difficulty": "hard"
  },
  "b2d0c4b4-ac68-491c-a561-60848ea34fb6": {
    "category": "Entertainment: Music",
    "answers": [
      "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
      "Night Fever",
      "The Chain",
      "Staying Alive"
    ],
    "correctAnswer": "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
    "id": "b2d0c4b4-ac68-491c-a561-60848ea34fb6",
    "question": "Madonna&#039;s song &quot;Hung Up&quot; includes a piece from which popular 70s song?",
    "difficulty": "medium"
  },
  "83cea789-1c6d-4757-a571-c79a14a9133f": {
    "category": "General Knowledge",
    "answers": ["Liver", "Skin", "Heart", "large Intestine"],
    "correctAnswer": "Skin",
    "id": "83cea789-1c6d-4757-a571-c79a14a9133f",
    "question": "What is the largest organ of the human body?",
    "difficulty": "easy"
  },
  "c051d919-2618-4f4d-bb42-5e13bdf58423": {
    "category": "General Knowledge",
    "answers": ["Vanilla", "Saffron", "Cinnamon", "Cardamom"],
    "correctAnswer": "Saffron",
    "id": "c051d919-2618-4f4d-bb42-5e13bdf58423",
    "question": "What is the world&#039;s most expensive spice by weight?",
    "difficulty": "medium"
  },
  "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8": {
    "category": "Entertainment: Video Games",
    "answers": ["Diamond/Pearl", "Ruby/Sapphire", "X/Y", "Black/White"],
    "correctAnswer": "Diamond/Pearl",
    "id": "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8",
    "question": "Which Pokemon generation did the fan-named &quot;Masuda Method&quot; first appear in? ",
    "difficulty": "easy"
  },
  "4db885c9-d4c1-4515-9e20-db71d6557cd6": {
    "category": "Entertainment: Video Games",
    "answers": [
      "Tateyuki Shigaraki",
      "Manfred Von Karma",
      "Gregory Edgeworth",
      "Bansai Ichiyanagi"
    ],
    "correctAnswer": "Bansai Ichiyanagi",
    "id": "4db885c9-d4c1-4515-9e20-db71d6557cd6",
    "question": "In the &quot;Ace Attorney&quot; series, who was the truly responsible of the forging of the autopsy report of the pivotal IS-7 incident?",
    "difficulty": "hard"
  },
}

// mock questions stats
export const mockDifficultyStats: TStat = {
  medium: 3,
  hard: 2,
  easy: 3,
  "all difficulties": 8
}

// mock category stats for different difficulties
export const mockCategoryStats: TStat = {
  "Entertainment: Film": 2,
  "Entertainment: Video Games": 2,
  "General Knowledge": 2,
  "Entertainment: Music": 2,
  "all categories": 8
}

export const mockEasyStats: TStat = {
  "Entertainment: Film": 1,
  "Entertainment: Video Games": 1,
  "General Knowledge": 1,
  "all categories": 3
}

export const mockHardStats: TStat = {
  "Entertainment: Music": 1,
  "Entertainment: Video Games": 1,
  "all categories": 2
}

export const mockMediumStats: TStat = {
  "Entertainment: Film": 1,
  "General Knowledge": 1,
  "Entertainment: Music": 1,
  "all categories": 3
}

// mock question
export const mockQuestion: IQuestion = {
  question: "Mock &lt;Question&gt;",
  id: "mockId",
  difficulty: "easy",
  correctAnswer: "Mock Correct Answer",
  category: "Mock Category",
  answers: [
    "Mocha &amp; Chai",
    "React &#124; Vue &#124; Angular",
    "Mock Aswer 3",
    "Mock Correct Answer",
  ],
};
