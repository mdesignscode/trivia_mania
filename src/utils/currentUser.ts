import { redirect, type Cookies } from "@sveltejs/kit";
import { validateSessionToken } from "./session";

// check if user is already logged in
export const isSignedIn = async (token: string | undefined, route: string) => {
        // don't redirect auth pages
        const isAuthRoute = ['/login', '/signup'].includes(route);
        if (!token && isAuthRoute) return;

        // user not logged in
        if (!token) throw redirect(302, '/login');

        const { user } = await validateSessionToken(token);
        // user signed in
        if (user) throw redirect(302, '/');

        // no user
        if (!user && isAuthRoute) return;
        else throw redirect(302, '/login');
}

export const getUser = async (cookies: Cookies) => {
        const token = cookies.get('session');
        if (!token) {
                throw redirect(302, '/login');
        }

        const { session, user } = await validateSessionToken(token);
        if (!session || !user) {
                throw redirect(302, '/login');
        }

        const userData = user.dataValues;
        delete userData.password;

        return {
                user: userData,
        };
}

