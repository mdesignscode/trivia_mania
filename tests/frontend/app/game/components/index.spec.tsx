import Question from "@/app/game/components";
import { mockQuestion } from "@/utils/mockData";
import { renderGameContext } from "@/utils/test_game_context";
import { screen } from "@/utils/test_utils";

describe("Question component", () => {
  it("Should render <RenderQuestion />", async () => {
    renderGameContext(<Question questionObj={mockQuestion} index={1} />);

    const container = await screen.findByTestId("question-container");
    expect(container).toBeInTheDocument();
  });

  it("snapshot matches", () => {
    // render component
    const { baseElement } = renderGameContext(<Question questionObj={mockQuestion} index={1} />);
    expect(baseElement).toMatchSnapshot();
  });
});
