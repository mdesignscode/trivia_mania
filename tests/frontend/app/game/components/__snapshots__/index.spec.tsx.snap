// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question component Answer buttons is not disabled if current question has already been answered 1`] = `
<body>
  <div>
    <div>
      <div
        class="sc-grXZZQ lakGlY question col gap-7 rounded-lg p-6"
        data-testid="question-container"
      >
        <div
          class="flex justify-between items-center"
        >
          <h1
            class="text-xl"
          >
            Mock Category
          </h1>
          <p
            class="text-xl"
            style="color: green;"
          >
            easy
          </p>
        </div>
        <h1
          class="text-2xl"
        >
          Mock &lt;Question&gt;
        </h1>
        <svg
          data-testid="timer-container"
          style="border: 2px solid green; border-radius: 5px;"
          viewBox="0 0 100 5"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            class="transition-all"
            stroke="green"
            stroke-width="20"
            x1="0"
            x2="0"
            y1="5"
            y2="5"
          />
          30
        </svg>
        <div
          class="question_options grid grid-cols-2 grid-rows-2 gap-4"
        >
          <span
            data-testid="Mocha & Chai"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mocha & Chai"
            >
              <span
                data-testid="feedback-incorrect-0"
              />
              <p>
                Mocha & Chai
              </p>
            </button>
          </span>
          <span
            data-testid="React | Vue | Angular"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="React | Vue | Angular"
            >
              <span
                data-testid="feedback-incorrect-1"
              />
              <p>
                React | Vue | Angular
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Aswer 3"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mock Aswer 3"
            >
              <span
                data-testid="feedback-incorrect-2"
              />
              <p>
                Mock Aswer 3
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Correct Answer"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mock Correct Answer"
            >
              <span
                data-testid="feedback-correct-3"
              />
              <p>
                Mock Correct Answer
              </p>
            </button>
          </span>
        </div>
        <span
          data-testid="continue-later-button"
        >
          <button
            class="sc-bcPKhP fVQoNj w-full"
          >
            Continue Later
          </button>
        </span>
        <audio
          id="success"
        >
          <source
            src="/success.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
        <audio
          id="error"
        >
          <source
            src="/error.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
      </div>
    </div>
  </div>
</body>
`;

exports[`Question component Disables answer buttons if current question has already been answered 1`] = `
<body>
  <div>
    <div>
      <div
        class="sc-grXZZQ lakGlY question col gap-7 rounded-lg p-6 sc-grXZZQ lakGlY"
        data-testid="question-container"
      >
        <div
          class="flex justify-between items-center"
        >
          <h1
            class="text-xl"
          >
            Mock Category
          </h1>
          <p
            class="text-xl"
            style="color: green;"
          >
            easy
          </p>
        </div>
        <h1
          class="text-2xl"
        >
          Mock &lt;Question&gt;
        </h1>
        <svg
          data-testid="timer-container"
          style="border: 2px solid green; border-radius: 5px;"
          viewBox="0 0 100 5"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            class="transition-all"
            stroke="green"
            stroke-width="20"
            x1="0"
            x2="0"
            y1="5"
            y2="5"
          />
          30
        </svg>
        <div
          class="question_options grid grid-cols-2 grid-rows-2 gap-4"
        >
          <span
            data-testid="Mocha & Chai"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              disabled=""
              id="Mocha & Chai"
            >
              <span
                data-testid="feedback-incorrect-0"
              >
                <svg
                  aria-hidden="true"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="30"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.75 9.75l4.5 4.5m0-4.5l-4.5 4.5M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
              <p>
                Mocha & Chai
              </p>
            </button>
          </span>
          <span
            data-testid="React | Vue | Angular"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              disabled=""
              id="React | Vue | Angular"
            >
              <span
                data-testid="feedback-incorrect-1"
              />
              <p>
                React | Vue | Angular
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Aswer 3"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              disabled=""
              id="Mock Aswer 3"
            >
              <span
                data-testid="feedback-incorrect-2"
              />
              <p>
                Mock Aswer 3
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Correct Answer"
          >
            <button
              class="sc-bcPKhP jLmHOz flex w-full justify-center gap-2 items-center animate__animated"
              disabled=""
              id="Mock Correct Answer"
              style="--animate-duration: 1s;"
            >
              <span
                data-testid="feedback-correct-3"
              >
                <svg
                  aria-hidden="true"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="30"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 12.75L11.25 15 15 9.75M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
              <p>
                Mock Correct Answer
              </p>
            </button>
          </span>
        </div>
        <span
          data-testid="continue-later-button"
        >
          <button
            class="sc-bcPKhP fVQoNj w-full"
          >
            Continue Later
          </button>
        </span>
        <span>
          <button
            class="sc-bcPKhP jLmHOz w-full"
          >
            Next Question
          </button>
        </span>
        <audio
          id="success"
        >
          <source
            src="/success.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
        <audio
          id="error"
        >
          <source
            src="/error.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
      </div>
    </div>
  </div>
</body>
`;

exports[`Question component snapshot matches 1`] = `
<body>
  <div>
    <div>
      <div
        class="sc-grXZZQ lakGlY question col gap-7 rounded-lg p-6"
        data-testid="question-container"
      >
        <div
          class="flex justify-between items-center"
        >
          <h1
            class="text-xl"
          >
            Mock Category
          </h1>
          <p
            class="text-xl"
            style="color: green;"
          >
            easy
          </p>
        </div>
        <h1
          class="text-2xl"
        >
          Mock &lt;Question&gt;
        </h1>
        <svg
          data-testid="timer-container"
          style="border: 2px solid green; border-radius: 5px;"
          viewBox="0 0 100 5"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            class="transition-all"
            stroke="green"
            stroke-width="20"
            x1="0"
            x2="0"
            y1="5"
            y2="5"
          />
          30
        </svg>
        <div
          class="question_options grid grid-cols-2 grid-rows-2 gap-4"
        >
          <span
            data-testid="Mocha & Chai"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mocha & Chai"
            >
              <span
                data-testid="feedback-incorrect-0"
              />
              <p>
                Mocha & Chai
              </p>
            </button>
          </span>
          <span
            data-testid="React | Vue | Angular"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="React | Vue | Angular"
            >
              <span
                data-testid="feedback-incorrect-1"
              />
              <p>
                React | Vue | Angular
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Aswer 3"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mock Aswer 3"
            >
              <span
                data-testid="feedback-incorrect-2"
              />
              <p>
                Mock Aswer 3
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Correct Answer"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mock Correct Answer"
            >
              <span
                data-testid="feedback-correct-3"
              />
              <p>
                Mock Correct Answer
              </p>
            </button>
          </span>
        </div>
        <span
          data-testid="continue-later-button"
        >
          <button
            class="sc-bcPKhP fVQoNj w-full"
          >
            Continue Later
          </button>
        </span>
        <audio
          id="success"
        >
          <source
            src="/success.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
        <audio
          id="error"
        >
          <source
            src="/error.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
      </div>
    </div>
  </div>
</body>
`;
